[Blaze] Validate campaign image dimensions

selanthiraiyan opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #12463
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Blaze backend expects the campaign image size to be greater than or equal to 500*500.

To inform users about this requirement, we are validating the selected campaign image dimensions and showing a local error alert about the expected image dimensions. 

Internal discussion in which we decided to follow the above approach - p1712745183124909/1712674079.063239-slack-C03L1NF1EA3

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

Prerequisites
- Woo Store eligible for Blaze
- A published product with an image with dimensions less than 500*500. 
- An image with dimensions less than 500*500 in your device/simulator library.

Steps
- Login into a Woo store eligible for Blaze.
- Navigate to the Products tab and tap on the product with the image with dimensions less than 500*500. 
- You should see the "Promote with Blaze" button.
- Tap "Promote with Blaze" -> "Confirm Details". 
- You should see an error alert asking you to select an image with minimum dimensions of 500*500.
- Tap "Edit ad" and navigate to the edit screen.
- Tap "Change image" and choose an image with dimensions less than 500*500. 
- You should see an error alert asking you to select an image with minimum dimensions of 500*500.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

#### Creation form - When a product image does not have 500 * 500

https://github.com/woocommerce/woocommerce-ios/assets/524475/6fa001b9-3719-48d8-b3e3-c12e239be92e

#### Edit image form - When selected image 500 * 500

https://github.com/woocommerce/woocommerce-ios/assets/524475/18ebacb4-68e6-4a2e-ac36-c26f7771ec1a


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

selanthiraiyan committed:
Remove unused values to fix warnings.

selanthiraiyan committed:
Show alert if image size is below 500 * 500.

selanthiraiyan committed:
Check that selected image is in expected dimensions.

selanthiraiyan committed:
Merge branch 'trunk' into feat/12463-blaze-image

selanthiraiyan committed:
Add release notes.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/194">18.2</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Blaze/BlazeEditAd/BlazeEditAdView.swift:
```diff
@@ -252,6 +255,16 @@ private extension BlazeEditAdView {
             )
+            static let imageSizeError = NSLocalizedString(
+                "blazeEditAdView.image.imageSizeError",
+                value: "Please select an image with minimum dimensions of 500 * 500.",
```
Nit: we should specify the unit for the image size for clarity.

```suggestion
                value: "Please select an image with minimum dimensions of 500 × 500 px.",
```

selanthiraiyan replied:
Updated in https://github.com/woocommerce/woocommerce-ios/pull/12470/commits/e0806346a0b0b918b91ec22e4d559076dc49cb41

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Blaze/BlazeEditAd/BlazeEditAdViewModel.swift:
```diff
@@ -154,6 +160,10 @@ final class BlazeEditAdViewModel: ObservableObject {
             guard let image = await onAddImage(source) else {
                 return imageState = previousState
             }
+            guard image.image.size.width >= minImageSize.width || image.image.size.height >= minImageSize.height else {
```
I'm not sure if checking the size is enough. I have an image of size 1722*501 but it still got rejected when creating a new campaign. In order to test this, you can open any image in Preview app > Tools > Adjust Size > then set either of the dimensions to 500 or 501. I haven't checked with slightly larger sizes, it'd be great if we can confirm the logic with backend folks.

selanthiraiyan replied:
I have updated the expected size as `600 * 600`. But we are still talking with the backend team as the campaign creation fails for image size `1722 * 601`. p1712848101999269/1712831053.533739-slack-C069RG07W6S

Meanwhile, since we started showing the exact error reason to the user, https://github.com/woocommerce/woocommerce-ios/pull/12470#issuecomment-2050215178 I think we can merge this PR for now and update the image size validation again.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Blaze/CampaignCreation/BlazeCampaignCreationFormViewModel.swift:
```diff
@@ -297,10 +298,14 @@ final class BlazeCampaignCreationFormViewModel: ObservableObject {
             return isShowingMissingImageErrorAlert = true
         }
 
+        guard image.image.size.width >= editAdViewModel.minImageSize.width || image.image.size.height >= editAdViewModel.minImageSize.height else {
```
Same comment as above about the check.

itsmeichigo reviewed (commented):
Thanks for the fix! I wasn't able to submit a campaign with an image of size 1722 × 501 - I'm not sure if our check matches with backend's check - I think we should verify with them. In the meanwhile, should we display a message when the image size is insufficient? We might want to catch this error:

> ⛔️ Error creating Blaze campaign: Sorry, there was an issue with the server. Please try again later. (Error code: 422). Response: {"message":"Validation failed","details":{"main_image":{"message":"Insufficient image size for cropping"}}}

selanthiraiyan committed:
Restore Blaze section in dashboard after campaign creation.

selanthiraiyan committed:
Validate product image dimensions before using it for campaign instead of showing error after tapping confirm button.

selanthiraiyan committed:
Merge branch 'trunk' into feat/12463-blaze-image

selanthiraiyan committed:
Update Blaze campaign image minimum size.

selanthiraiyan committed:
Merge branch 'trunk' into feat/12463-blaze-image

selanthiraiyan committed:
Handle insufficient image size error and show error message.

selanthiraiyan committed:
Update error message.

selanthiraiyan commented:
> In the meanwhile, should we display a message when the image size is insufficient? We might want to catch this error:

Great point. I have started handling this error in https://github.com/woocommerce/woocommerce-ios/pull/12470/commits/73268afefd5c29fa30773f1dca36acc2674e49ee

#### When the insufficient error happens
| Before | After |
|--------|--------|
| ![Before](https://github.com/woocommerce/woocommerce-ios/assets/524475/956dad8e-47db-4a70-acfb-d90fc92ca35e) | ![After](https://github.com/woocommerce/woocommerce-ios/assets/524475/0c978f91-f4c1-41d6-abcc-b42b555517cb) |

selanthiraiyan commented:
Thank you for the review, @itsmeichigo! 🙇 The PR is ready for another look. 

As discussed in p1712739200004669-slack-C03L1NF1EA3, I have added the change of resetting the My Store Blaze section after campaign creation in https://github.com/woocommerce/woocommerce-ios/pull/12470/commits/e4560e7e7a74305b31e5249c8b621e0333162583. 

I have addressed your comments. We still aren't sure about the expected image dimensions, but we have started showing a better error message. I think we can merge this PR as a temporary fix and update the image size later.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Blaze/BlazeEditAd/BlazeEditAdView.swift:
```diff
@@ -252,6 +255,16 @@ private extension BlazeEditAdView {
             )
+            static let imageSizeError = NSLocalizedString(
+                "blazeEditAdView.image.imageSizeError",
+                value: "Please select an image with minimum dimensions of 600 * 600 px.",
```
We should use "× " instead of "*" for user-facing copy (the same symbol is used on mac).

```suggestion
                value: "Please select an image with minimum dimensions of 600 × 600 px.",
```

itsmeichigo replied:
I'll merge the change above and enable automerge for the code freeze.

selanthiraiyan replied:
Thank you! ❤️

itsmeichigo reviewed (approved):
I tested again with an image of size 2062 × 600 and successfully created a campaign 🎉  Resetting Blaze section visibility works too. 🚢

itsmeichigo committed:
Update WooCommerce/Classes/ViewRelated/Blaze/BlazeEditAd/BlazeEditAdView.swift
